### PR TITLE
[#5008] More granular control over tools/panels

### LINKED ISSUE #5008: More granular control over tools/panels
It would be nice to have the ability to remove unnecessary controls and, for example, leave expand to full screen, etc. It can be done by passing either object with an explicit state of each available control element(`{zoom: true, workbench: false}`) or by choosing one of a predefined set of controls(like current `hideExplorerPanel=1`, but extended to `toolset=standard-no-explorer`), which is less flexible but, likely, easier to implement

### PR DESCRIPTION
**Pull request by smotornyuk**
### Add ability to hide different buttons\panels using init files

Fixes #5008

This PR adds a new wrapper `visibilitySwitch` that conditionally renders elements depending on their configuration from init-file. Inside init-file, user can add new `elements` sections that looks like the following: 
```json
  "elements": {
    "show-workbench": {"visible": false},
    "map-data-count": {"visible": false},
    "help": {"visible": false},
    "menu-bar": {"visible": false},
    "map-navigation": {"visible": true},
    "compass": {"visible": false},
    "zoom": {"visible": true},
    "my-location": {"visible": false},
    "split-tool": {"visible": false},
    "about-link": {"visible": false},
    "measure-tool": {"visible": false},
    "feedback": {"visible": false},
    "related-maps": {"visible": false},
    ...
  },
```

Later these config objects are passed by-name into corresponding elements(buttons, panels), and depending on the `visible` flag they either rendered or not. This is a rather hot-fix and I want to provide a more elegant solution, but it requires more essential changes. Here a few points that are not completely clear to me:

* [Configuration for elements](https://github.com/TerriaJS/terriajs/compare/next...DataShades:5008-elements-visibility?expand=1#diff-4095bc5d59253a1e3670aa8c6391ceacc1b33a385263394c2eeb12a8737a95d3R180) stored inside `Terria` instance. How to make it reactive, so that it updates every time user changes init file(for example, retypes `#<INIT>` part of the URL). Right now only initial values, that were available at load time are applied
* From my point of view, it has the sense to use the same common React component for all of the buttons\controls and pass different props\children in order to style those buttons. Right now only a few elements are using `MapIconButton`, all others are using a unique way of rendering.
* Maybe we can pass `terria/viewState` config object through context so that it will be possible to access it using hooks inside any component?
* Where is the best place to put `visibilitySwitch` from the conceptual point of view? Maybe, `HOCs`?
* Is there any sense in collecting all the names of elements with controlled visibility into some kind of service? For example, developers can use this service for listing all the available names and(potentially) for switching states of every separate component in real-time.

### Checklist

-   [ ] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
-   [ ] I've updated CHANGES.md with what I changed.

### CODE REVIEW

**Comment by tephenavies** [comment]
Hey @smotornyuk. I really like the idea of this PR. I'll do a full code review of this hopefully tomorrow.

In response to your questions:
1. Element config reactivity: I'll have to look further into the code to answer this
2. Inconsistency in drawing of buttons: I don't know of any reasons not to use MapIconButtons for everything. It's likely that buttons that have been modified recently use the new MapIconButton and others don't and should be updated.
3. Passing terria/viewState through React context and access with hooks: We should do this. It will slightly change how e.g. tests are written (components in jsx/tsx will have to be wrapped in the context provider when not rendered inside the Terria UI) but I think it'll make developing the UI easier.
4. Where should `visibilitySwitch` go? `HOCs` is the right place.
5. Where should we list all of the elements that can be configured? At the moment I think we should add a page to our docs https://docs-v8.terria.io about the functionality and list the element names there. (PRs to `mobx-docs` branch for updated docs would be appreciated. I'll try to get that merged onto `next` soon too).

**Review by tephenavies — CHANGES_REQUESTED** [review]
Hi @smotornyuk. Thanks for the PR. I think `Terria.elements` is only not reactive due to limits of MobX v4 (see the dot point about MobX 4 in [MobX `object` docs](https://github.com/mobxjs/mobx/blob/4.15.7/docs/refguide/object.md) if you're interested). It should be reactive if you switch to using an `ObservableMap`.
Other than that I've added some suggestions for small improvements to change/discuss.

**Comment by zoran995** [comment]
Hey, @steve9164 we also like the idea of introducing more controls of navigation items but I am still more into introducing the navigation model component that would control all navigation items. It would probably reduce the complexity of the component rendering, and also force us to have all of the controls for the new navigation item.

The one written in #5062 is just a quickly written prototype for a client on how it could look and work (the basic idea is taken from vs code models), and I am aware that some parts of it should go through a rewrite and extended on menu bar. The nav model also might be connected to Stratum and Traits system so we benefit from multiple layers of configurability. We should probably separate it into two parts one just to introduce the model, and then grouping of items on smaller screens. If you agree with proposed and @smotornyuk is interested and have time I will be happy to let him take it over.

**Comment by tephenavies** [comment]
Hey @zoran995. I think you're right. It would be beneficial to have the navigation controls framework reviewed before merging this. I'll give it a review this week.

File: lib/Models/Terria.ts [hunk 1]
```diff
@@ -175,6 +176,9 @@ export default class Terria {
   readonly timelineClock = new Clock({ shouldAnimate: false });
   // readonly overrides: any = overrides; // TODO: add options.functionOverrides like in master
 
+  @observable
+  readonly elements = new ElementsConfig();
```
[the remark this answers is not in the document]
Comment (smotornyuk) [reply lib/Models/Terria.ts]: Yep, this way it works as expected

**Comment by smotornyuk** [comment]
@steve9164, thank you for review. `observable.map` really fixed the thing:) And I've applied other changes 

I'm not sure [about this place](https://github.com/TerriaJS/terriajs/pull/5131/files#diff-1a7fa518f4480d4e75616da75517223c7e6c07f5b971727c8878550125831f86R22) - i had to extend type with `any` because otherwise, I'm constantly getting the following error and solutions from StackOverflow don't look nice either
```
ERROR in packages/terriajs/lib/ReactViews/HOCs/withControlledVisibility.tsx:19:57
TS2700: Rest types may only be created from object types.
    17 | ) => {
    18 |   // eslint-disable-next-line require-jsdoc
  > 19 |   function WithControlledVisibility({ elementConfig, ...props }: P & WithControlledVisibilityProps) {
       |                                                         ^^^^^
    20 |     const isVisible = elementConfig ? elementConfig.visible : true;
    21 |     return isVisible ? <WrappedComponent {...props} /> : null;
    22 |   }
```

**Comment by AnaBelgun** [comment]
@steve9164 this will need to be checked and merged by 12-13 April max; the dashboard depending on it will be launched mid April and it needs a couple of days to test the data etc etc

**Pull request merged**